### PR TITLE
Refactor weapon_medkit

### PR DESCRIPTION
**Pull request by Kefta**
- Converted the ammo/idle timers to predicted netvars like weapon_fists
- Divided up the Primary/SecondaryAttack functions into SWEP:CanHeal(entity), SWEP:DoHeal(entity), SWEP:HealEntity(entity), and SWEP:HealFail(entity)
- Changed many constants to class variables to promote addon SWEP modification instead of copying and modifying the code
- Heal success/failures now punishes both primary and secondary fire no matter which heal type it is to prevent being able to interrupt the medpack animation with the opposite heal type attack
- Regen time is now tied to when the last heal was performed or when the medkit was deployed instead of being arbitrarily tied to a timer's creation time
- Removed SWEP.MaxAmmo as SWEP.Primary.ClipSize was already serving its purpose
- General optimisations/cleanup

Supersedes https://github.com/Facepunch/garrysmod/pull/1942.

### CODE REVIEW

**Comment by Kefta** [comment]
Tested in local multiplayer and works without regressions as far as I can tell other than the changed behaviour noted.

**Comment by Kefta** [comment]
Can this please be reviewed? There's really not that many built-in weapons and it would be good to have at least one example of SWEP design that takes runtime modification into account

**Review by robotboy655 — APPROVED** [review]

File: garrysmod/lua/weapons/weapon_medkit.lua [hunk 1]
```diff
@@ -18,135 +18,178 @@ SWEP.UseHands = true
...
+	if ( isplayer ) then
 		owner:LagCompensation( true )
 	end
 
+	local startpos = owner:GetShootPos()
 	local tr = util.TraceLine( {
-		start = owner:GetShootPos(),
-		endpos = owner:GetShootPos() + owner:GetAimVector() * 64,
+		start = startpos,
+		endpos = startpos + owner:GetAimVector() * 64,
```
Comment (2048khz-gachi-rmx) [inline garrysmod/lua/weapons/weapon_medkit.lua:74]: I believe `self.HealRange` was intended to be used here

Comment (Kefta) [reply garrysmod/lua/weapons/weapon_medkit.lua:74]: You're right

**Pull request merged**